crypto/fipsmodule/rand/new_rand.c

```diff
@@ -0,0 +1,290 @@
+    }
+    first_pred_resistance_len = RAND_PRED_RESISTANCE_LEN;
+  }
+
+  assert(first_pred_resistance_len == 0 ||
+         first_pred_resistance_len == RAND_PRED_RESISTANCE_LEN);
+
+  // Iterate CTR-DRBG generate until we generated |out_len| bytes of randomness.
+  while (out_len > 0) {
+    size_t todo = out_len;
+    if (todo > CTR_DRBG_MAX_GENERATE_LENGTH) {
+      todo = CTR_DRBG_MAX_GENERATE_LENGTH;
+    }
+
+    // Each reseed interval can generate up to
+    // |CTR_DRBG_MAX_GENERATE_LENGTH*2^{kCtrDrbgReseedInterval}| bytes.
+    // Determining the time(s) to reseed prior to entering the CTR-DRBG generate
+    // loop is a doable strategy. But tracking reseed times adds unnecessary
+    // complexity. Instead our strategy is optimizing for simplicity.
+    // |out_len < CTR_DRBG_MAX_GENERATE_LENGTH| will be the majority case
+    // (by far) and requires a single check in either strategy.
+    // Note if we just reseeded through |rand_ctr_drbg_reseed()|, we won't
+    // reseed again here.
+    if( state->generate_calls_since_seed + 1 >= kCtrDrbgReseedInterval) {
+      rand_ctr_drbg_reseed(state);
+    }
+
+    if (!CTR_DRBG_generate(&state->drbg, out, todo, pred_resistance,
+          first_pred_resistance_len)) {
+      abort();
+    }
+
+    out += todo;
+    out_len -= todo;
+    state->generate_calls_since_seed++;
+    first_pred_resistance_len = 0;
+  }
+
+  OPENSSL_cleanse(pred_resistance, RAND_PRED_RESISTANCE_LEN);
+
+  if (rand_ensure_valid_state() != 1) {
+    abort();
+  }
+}
+
+static void RAND_bytes_private(uint8_t *out, size_t out_len,
+  const uint8_t user_pred_resistance[RAND_PRED_RESISTANCE_LEN],
+  int use_user_pred_resistance) {
+
+  if (out_len == 0) {
+    return;
+  }
+
+  struct rand_thread_local_state *state =
+      CRYPTO_get_thread_local(OPENSSL_THREAD_LOCAL_PRIVATE_RAND);
+
+  if (state == NULL) {
+    state = OPENSSL_zalloc(sizeof(struct rand_thread_local_state));
+    if (state == NULL ||
+        CRYPTO_set_thread_local(OPENSSL_THREAD_LOCAL_PRIVATE_RAND, state,
+                                   rand_thread_local_state_free) != 1) {
+      abort();
+    }
+
+    rand_state_initialize(state);
+  }
+
+  RAND_bytes_core(state, out, out_len, user_pred_resistance,
+    use_user_pred_resistance);
+}
+
+// TOOD
+// Retire and replace call sites with RAND_bytes_with_user_prediction_resistance
+int NR_PREFIX(RAND_bytes_with_additional_data)(uint8_t *out, size_t out_len,
+  const uint8_t user_pred_resistance[RAND_PRED_RESISTANCE_LEN]) {
+  
+  RAND_bytes_private(out, out_len, user_pred_resistance,
+    RAND_USE_USER_PRED_RESISTANCE);
+  return 1;
+}
+
+int RAND_bytes_with_user_prediction_resistance(uint8_t *out, size_t out_len,
+  const uint8_t user_pred_resistance[RAND_PRED_RESISTANCE_LEN]) {
+  
+  RAND_bytes_private(out, out_len, user_pred_resistance,
+    RAND_USE_USER_PRED_RESISTANCE);
+  return 1;
+}
+
+int NR_PREFIX(RAND_bytes)(uint8_t *out, size_t out_len) {
```

Comment:
Shouldn't this use a different "public" random bytes? If everything uses the "private" one it's not exactly private.

Reply:
I'm still pondering. Later I want to add another, independent state, such that we can have a real public/private API. But which exported function should have the "private" state and which should have the "public" state? You can view it two ways:

1. Every output of `RAND_bytes` is assumed to come from "private". You have to actively select you want some public output.
2. Every output of `RAND_bytes` is assumed to be "public". 

I am probably going for (2). I will have to rename some stuff later. But that is just renaming some internal functions.

Firstly, let's say we pick (2). Then if an application want to catagorise into public/private, then they need to change all function calls that generate private output. You can miss some call sites. Similarly, if we pick (1), you can still miss call sites when migrating; it's just the reverse where you forgot some call sites that should have been made public. It's sorta equivalent.

Secondly, everyone are already using `RAND_bytes()` in their code. We are not going to change this API, ever. Hence, to avoid having to add yet another function "RAND_bytes_public"-something, we can make use of `RAND_priv_bytes`.

In fact, it's better not to think about this as public/private. A better way to think about it, conceptually, is that you have two independent states from which you can generate randomness. Applications can then do the catagorisation they want. But since public/private is already used. We are stuck with that naming.

Reply:
Eagerly awaiting this change, have you talked to any of our customers (such as s2n-tls) what they would expect or prefer?

Reply:
`RAND_bytes` and `RAND_priv_bytes` exists in OpenSSL and are not deprecated. So, I think it would make most sense to stick to those. They are pretty natural to use for e.g. s2n-tls.

Reply:
Makes sense to use those names, I was thinking about the choice between open 1 and 2, and what (if anything) in AWS-LC's own use of RAND_bytes should be migrated to RAND_priv_bytes.

Reply:
> (if anything) in AWS-LC's own use of RAND_bytes should be migrated to RAND_priv_bytes.

We can definitely look at that later.